### PR TITLE
Feat: profiles moved from vault commands

### PR DESCRIPTION
**Pull request by burdandrei**
- Profiles were redone to cache ephemeral tokens
- Use GitHub for Vault auth
- Use Vault got Consul Auth
- Use Vault for Nomad Auth

### CODE REVIEW

File: command/profile/use_profile.go [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+					et, err := time.Parse(time.RFC3339Nano, profileCache.Vault.Auth.ExpireTime)
+					if err != nil { // if expire time  can't be parsed as time assume cache is bad and relogin
+						profileCache.Vault, err = vaultLoginGitHub(profile.Vault, v)
+						if err != nil {
+							fmt.Printf("Bad github login for profile %s", name)
+						}
+					} else if et.Before(time.Now()) { // relogin if token is expired
+						profileCache.Vault, err = vaultLoginGitHub(profile.Vault, v)
+						if err != nil {
+							fmt.Printf("Bad github login for profile %s", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: I find it confusing that you mix outputting `export ...` with error messages

File: command/profile/use_profile.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	if profile.Consul.Auth.Method == "vault" {
+		if profileCache.Consul.Auth.Token != "" {
+			if profileCache.Consul.Auth.ExpireTime != "" {
+				et, err := time.Parse(time.RFC3339Nano, profileCache.Consul.Auth.ExpireTime)
+				if err != nil { // if expire time  can't be parsed as time assume cache is bad and relogin
+					profileCache.Consul, err = vaultGetConsulCreds(profile.Consul, v)
+					if err != nil {
+						fmt.Errorf("Error reading Consul creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+					}
+				} else if et.Before(time.Now()) { // relogin if token is expired
+					profileCache.Consul, err = vaultGetConsulCreds(profile.Consul, v)
+					if err != nil {
+						fmt.Errorf("Error reading Consul creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+					}
+				}
+			}
+		} else { // token  is absent cache - login
+			profileCache.Nomad, err = vaultGetNomadCreds(profile.Nomad, v)
+			if err != nil {
+				fmt.Errorf("Error reading Consul creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+			}
+		}
+		fmt.Printf("export CONSUL_HTTP_TOKEN=%)\n", profileCache.Consul.Auth.Token)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: should it be `%s` instead of `%`?

File: command/profile/use_profile.go [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	if profile.Nomad.Auth.Method == "vault" {
+		if profileCache.Nomad.Auth.Token != "" {
+			if profileCache.Nomad.Auth.ExpireTime != "" {
+				et, err := time.Parse(time.RFC3339Nano, profileCache.Nomad.Auth.ExpireTime)
+				if err != nil { // if expire time  can't be parsed as time assume cache is bad and relogin
+					profileCache.Nomad, err = vaultGetNomadCreds(profile.Nomad, v)
+					if err != nil {
+						fmt.Errorf("Bad reading of nomad creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+					}
+				} else if et.Before(time.Now()) { // relogin if token is expired
+					profileCache.Nomad, err = vaultGetNomadCreds(profile.Nomad, v)
+					if err != nil {
+						fmt.Errorf("Bad reading of nomad creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+					}
+				}
+			}
+		} else { // token  is absent cache - login
+			profileCache.Nomad, err = vaultGetNomadCreds(profile.Nomad, v)
+			if err != nil {
+				fmt.Errorf("Bad reading of nomad creds for %s profile ", name)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+		}
+		fmt.Printf("export NOMAD_TOKEN=%s\n", profileCache.Nomad.Auth.Token)
+	}
+
+	profilesCache[name] = profileCache
+
+	// Create a file for cache update
+	cacheTempFile, err := ioutil.TempFile("", "hashi_helper_cache")
+	if err != nil {
+		panic(err)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: why `panic` instead of returning the error?

File: command/profile/use_profile.go [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	yamlReadableCache, err := yaml.Marshal(&profilesCache)
+	if err != nil {
+		fmt.Errorf("Cache generation failed")
```
Comment (lorenzo) [inline command/profile/use_profile.go]: This error is going nowhere

File: command/profile/use_profile.go [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	// Write to the file
+	if err := ioutil.WriteFile(cacheTempFile.Name(), yamlReadableCache, 600); err != nil {
+		panic(err)
```
Comment (lorenzo) [inline command/profile/use_profile.go]: why `panic?`

Comment (burdandrei) [reply command/profile/use_profile.go]: this was copy paste

File: command/profile/use_profile.go (hunk 2 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 3 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 4 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 8 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 9 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 11 above)
Comment (burdandrei) [reply command/profile/use_profile.go]: fixed

File: command/profile/use_profile.go (hunk 9 above)
Comment (lorenzo) [reply command/profile/use_profile.go]: Not really, it still says `panic`

**Pull request merged**